pkg/cloudprovider/providers/openstack/openstack.go

```diff
@@ -294,19 +306,47 @@ func setConfigFromSecret(cfg *Config) error {
 		return err
 	}
 
+	cfg := &Config{}
 	if content, ok := secret.Data["clouds.conf"]; ok {
 		err = gcfg.ReadStringInto(cfg, string(content))
 		if err != nil {
 			klog.Errorf("Cannot parse data from the secret.")
 			return fmt.Errorf("cannot parse data from the secret")
 		}
+		newOpenStack, err := newOpenStack(*cfg)
+		if err != nil {
+			klog.Errorf("Cannot initialize cloud provider using data from the secret.")
+			return fmt.Errorf("cannot initialize cloud provider using data from the secret")
+		}
+		os.provider = newOpenStack.provider
+		os.region = cfg.Global.Region
+		os.lbOpts = cfg.LoadBalancer
+		os.bsOpts = cfg.BlockStorage
+		os.routeOpts = cfg.Route
+		os.metadataOpts = cfg.Metadata
+		klog.Info("OpenStack cloud provider was initialized.")
 		return nil
 	}
 
 	klog.Errorf("Cannot find \"clouds.conf\" key in the secret.")
 	return fmt.Errorf("cannot find \"clouds.conf\" key in the secret")
 }
 
+func (os *OpenStack) ensureCloudProviderWasInitialized() error {
```

Comment:
Should we cache the result of this function?

Reply:
i think the result is "cached" by setting `os.provider` when the initialization succeeds (checked first thing in the next line).  correct me if i'm wrong.

Reply:
Yes :+1: